sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -463,6 +463,8 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
       case If(Literal(null, _), _, falseValue) => falseValue
       case If(cond, trueValue, falseValue)
         if cond.deterministic && trueValue.semanticEquals(falseValue) => trueValue
+      case If(cond, l @ Literal(null, _), FalseLiteral) if !cond.nullable => And(cond, l)
+      case If(cond, l @ Literal(null, _), TrueLiteral) if !cond.nullable => Or(Not(cond), l)
```

Comment:
`cond.deterministic` seems to be required, doesn't it?

Reply:
For a Spark `Expression`, deterministic means that for fixed input it should always return the same result (e.g., a deterministic function), while `nullable` means whether it will be evaluated to null. In this case we need the latter.

Reply:
@sunchao I think the concern @dongjoon-hyun has is it's possible that `cond` is non-deterministic or has side-effect (all the expressions that have side effects in Spark, we consider them as non-deterministic). If we naively simplify them, we might change the final result. 

For example, `If(cond, TrueLiteral, TrueLiteral)` can be always simplified to `TrueLiteral`; however, this will skip evaluating `cond` which might have side-effect such as counting the number of times we call `cond`. 

In the first case, the `cond` in `And(cond, l)` will be alway evaluated, and in the second case, `Or(Not(cond), l)`, since `l` will never be `true`, `Not(cond)` will always be evaluated.

As a result, maybe we don't need such strong requirement that `cond` has to be deterministic.

Reply:
FYI,
```
      case If(cond, trueValue, falseValue)
        if cond.deterministic && trueValue.semanticEquals(falseValue) => trueValue
```
This rule is an example why `cond` has to be deterministic.

Reply:
In the added patterns, `cond` isn't skipped or changed with evaluation order, I think `cond.deterministic` shouldn't be required.

Reply:
I see. Thanks @dbtsai @dongjoon-hyun and @viirya , yeah I think in this case we don't require `cond` to be deterministic.